Unhandled error with variadic functions

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
fn five() -> i8 {
    5
}

extern "C" {
    fn printf(fmt: *const i8, ...);
}

fn main() {
    let a = 127i8;
    let b = five();

    let c = a + b;

    unsafe { printf("%d\n\0" as *const str as *const i8, c) }
}

```

I expected to see this happen: `rustc` errors out with the following:
```rust
error[E0617]: can't pass `i8` to variadic function
  --> test.rs:15:58
   |
15 |     unsafe { printf("%d\n\0" as *const str as *const i8, c) }
   |                                                          ^ help: cast the value to `c_int`: `c as c_int`
```

Instead, this happened: no error

--- a/gcc/rust/typecheck/rust-tyty-call.cc
+++ b/gcc/rust/typecheck/rust-tyty-call.cc
@@ -122,7 +122,7 @@ TypeCheckCallExpr::visit (FnType &type)
 	  return;
 	}
 
-      // it might be a varadic function
+      // it might be a variadic function
       if (i < type.num_params ())
 	{
 	  auto fnparam = type.param_at (i);
@@ -141,6 +141,60 @@ TypeCheckCallExpr::visit (FnType &type)
 	  if (resolved_argument_type->get_kind () == TyTy::TypeKind::ERROR)
 	    {
 	      return;
+	    }
+	}
+      else
+	{
+	  switch (argument_expr_tyty->get_kind ())
+	    {
+	    case TyTy::TypeKind::ERROR:
+	      return;
+	      case TyTy::TypeKind::INT: {
+		auto &int_ty
+		  = static_cast<TyTy::IntType &> (*argument_expr_tyty);
+		if ((int_ty.get_int_kind () == TyTy::IntType::IntKind::I8)
+		    || (int_ty.get_int_kind () == TyTy::IntType::IntKind::I16))
+		  {
+		    rust_error_at (arg_locus,
+				   "expected %<c_int%> variadic argument");
+		    return;
+		  }
+		break;
+	      }
+	      case TyTy::TypeKind::UINT: {
+		auto &uint_ty
+		  = static_cast<TyTy::UintType &> (*argument_expr_tyty);
+		if ((uint_ty.get_uint_kind () == TyTy::UintType::UintKind::U8)
+		    || (uint_ty.get_uint_kind ()
+			== TyTy::UintType::UintKind::U16))
+		  {
+		    rust_error_at (arg_locus,
+				   "expected %<c_uint%> variadic argument");
+		    return;
+		  }
+		break;
+	      }
+	      case TyTy::TypeKind::FLOAT: {
+		if (static_cast<TyTy::FloatType &> (*argument_expr_tyty)
+		      .get_float_kind ()
+		    == TyTy::FloatType::FloatKind::F32)
+		  {
+		    rust_error_at (arg_locus,
+				   "expected %<c_double%> variadic argument");
+		    return;
+		  }
+		break;
+	      }
+	    case TyTy::TypeKind::BOOL:
+	      rust_error_at (arg_locus, "expected %<c_int%> variadic argument");
+	      return;
+	    case TyTy::TypeKind::FNDEF:
+	      rust_error_at (arg_locus,
+			     "unexpected function definition type as variadic "
+			     "argument - cast to function pointer");
+	      return;
+	    default:
+	      break;
 	    }
 	}
 

--- a/gcc/testsuite/rust/compile/variadic.rs
+++ b/gcc/testsuite/rust/compile/variadic.rs
@@ -1,0 +1,8 @@
+extern "C" {
+    fn printf(s: *const i8, ...);
+}
+
+fn main() {
+    // { dg-error "expected .c_int. variadic argument" "" { target *-*-* } .+1 }
+    printf("%d\n" as *const str as *const i8, 1i8);
+}

--- a/gcc/testsuite/rust/execute/torture/overflow1.rs
+++ b/gcc/testsuite/rust/execute/torture/overflow1.rs
@@ -16,5 +16,5 @@ fn main() {
     // { dg-final { scan-tree-dump ADD_OVERFLOW original } }
     let c = a + b;
 
-    unsafe { printf("%d\n\0" as *const str as *const i8, c) }
+    unsafe { printf("%d\n\0" as *const str as *const i8, c as i32) }
 }

